### PR TITLE
Adds cloud-console integration

### PR DESCRIPTION
**Pull request by ashraffouda**
Signed-off-by: Ashraf Fouda <ashraf.m.fouda@gmail.com>

### CODE REVIEW

File: pkg/vm/ch.go [hunk 1]
```diff
@@ -17,9 +19,34 @@ import (
 )
 
 const (
-	chBin = "cloud-hypervisor"
+	chBin           = "cloud-hypervisor"
+	cloudConsoleBin = "cloud-console"
 )
 
+func (m *Machine) StartCloudConsole(ctx context.Context, namespace string, networkAddr net.IPNet, machineIP net.IPNet, ptyPath string) error {
```
Comment (xmonader) [inline pkg/vm/ch.go]: Exported function missing a comment

File: pkg/vm/ch.go [hunk 2]
```diff
@@ -17,9 +19,34 @@ import (
 )
 
 const (
-	chBin = "cloud-hypervisor"
+	chBin           = "cloud-hypervisor"
+	cloudConsoleBin = "cloud-console"
```
Comment (xmonader) [inline pkg/vm/ch.go:26]: where does this binary come from? is there a PR for that bin?

File: pkg/vm/ch.go [hunk 3]
```diff
@@ -17,9 +19,34 @@ import (
...
 )
 
+func (m *Machine) StartCloudConsole(ctx context.Context, namespace string, networkAddr net.IPNet, machineIP net.IPNet, ptyPath string) error {
+	machineIPLsit := strings.Split(machineIP.IP.String(), ".")
```
Comment (xmonader) [inline pkg/vm/ch.go]: machineIPLsit is a typo, not sure if you mean the ipOctets that are separated by dots if so please use ipOctets instead?

File: pkg/vm/ch.go [hunk 4]
```diff
@@ -17,9 +19,34 @@ import (
...
+	port, err := strconv.Atoi(machineIPLsit[len(machineIPLsit)-1])
+	if err != nil {
+		return err
+	}
+	port = 20000 + port
```
Comment (xmonader) [inline pkg/vm/ch.go]: where does the 20000 number come from? and is there a guard against max port used 65535?

File: pkg/vm/ch.go [hunk 5]
```diff
@@ -17,9 +19,34 @@ import (
...
+	args := []string{"setsid", "ip", "netns", "exec", namespace, cloudConsoleBin, ptyPath, networkAddr.IP.String(), strconv.Itoa(port)}
+	log.Debug().Msgf("running cloud-console : %+v", args)
+
+	cmd := exec.CommandContext(ctx, "busybox", args...)
+
+	if err := cmd.Start(); err != nil {
+		return errors.Wrap(err, "failed to start cloud-hypervisor")
+	}
+
+	if err := m.release(cmd.Process); err != nil {
```
Comment (xmonader) [inline pkg/vm/ch.go:58]: you can do return `m.release(cmd.Process)` directly

File: pkg/vm/ch.go [hunk 6]
```diff
@@ -33,7 +60,7 @@ func (m *Machine) Run(ctx context.Context, socket, logs string) error {
 		"--memory": {fmt.Sprintf("%s,shared=on", m.Config.Mem.String())},
 
 		"--console":    {"off"},
-		"--serial":     {"tty"},
+		"--serial":     {"pty"},
```
Comment (xmonader) [inline pkg/vm/ch.go]: can you please add a comment here why the change to pty

File: pkg/vm/client.go [hunk 7]
```diff
@@ -15,6 +15,12 @@ type Client struct {
 	client http.Client
 }
 
+type VMData struct {
+	CPUs    CPU
```
Comment (xmonader) [inline pkg/vm/client.go]: We should use CPU or NumCPU instead of CPUs

File: pkg/vm/client.go [hunk 8]
```diff
@@ -86,22 +92,22 @@ func (c *Client) Resume(ctx context.Context) error {
 }
 
 // Inspect return information about the vm
-func (c *Client) Inspect(ctx context.Context) (CPU, MemMib, error) {
+func (c *Client) Inspect(ctx context.Context) (VMData, error) {
```
Comment (xmonader) [inline pkg/vm/client.go:96]: Not sure if Inspect is better than Info -specially we are calling vm.info :) - , but if Inspect is used everywhere we can be consistent

File: pkg/vm/client.go [hunk 9]
```diff
@@ -86,22 +92,22 @@ func (c *Client) Resume(ctx context.Context) error {
...
 
 	response, err := c.client.Do(request)
 	if err != nil {
-		return 0, 0, errors.Wrap(err, "error calling machine info")
+		return VMData{}, errors.Wrap(err, "error calling machine info")
 	}
 
 	defer response.Body.Close()
 
 	if response.StatusCode != http.StatusOK {
```
Comment (xmonader) [inline pkg/vm/client.go:110]: is it possible to have a body that we can also add in the error message along with the status code?

File: pkg/vm/manager.go [hunk 10]
```diff
@@ -471,8 +472,34 @@ func (m *Module) Run(vm pkg.VM) error {
...
+			return errors.Wrapf(err, "failed to get network '%s'", ifc.NetID)
+		}
+
+		client := NewClient(m.socketPath(vm.Name))
+		vmData, err := client.Inspect(ctx)
+		if err != nil {
+			return errors.Wrapf(err, "failed to Inspect vm with id '%s'", vm.Name)
+		}
+
+		err = machine.StartCloudConsole(ctx, namespace, networkAddr, ifc.IPs[0], vmData.PTYPath)
```
Comment (xmonader) [inline pkg/vm/manager.go]: can you please add a comment on ifc.IPs[0] usage

File: pkg/vm/manager.go [hunk 11]
```diff
@@ -471,8 +472,34 @@ func (m *Module) Run(vm pkg.VM) error {
...
+		if err != nil {
+			return errors.Wrapf(err, "failed to start cloud-console for vm id '%s'", vm.Name)
+		}
+		return nil
```
Comment (xmonader) [inline pkg/vm/manager.go]: is this meant to do an early return from the loop as soon as you started the cloud console?

**Comment by xmonader** [comment]
I added some minor comments and questions, also please make sure to add docs file that explains the cloud console flow

**Review by xmonader — CHANGES_REQUESTED** [review]
please check the comments

File: pkg/vm/manager.go (hunk 11 above)
Comment (ashraffouda) [reply pkg/vm/manager.go]: yes

File: docs/manual/zmachine/cloud-console.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# Cloud console
+
+- `cloud-console` is a tool to view machine logging and interact with the machine you have deployed
+- It always runs on the machine's private network ip and port number equla to 20000+last octect of machine private IP
```
Comment (xmonader) [inline docs/manual/zmachine/cloud-console.md]: equla -> equals

File: pkg/vm/ch.go (hunk 2 above)
Comment (xmonader) [reply pkg/vm/ch.go:26]: resolved in bins/packages/cloudconsole/cloudconsole.sh

File: pkg/vm/ch.go (hunk 1 above)
Comment (xmonader) [reply pkg/vm/ch.go]: the comment should start with the method name `StartCloudConsole`

File: docs/manual/zmachine/cloud-console.md [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# Cloud console
+
+- `cloud-console` is a tool to view machine logging and interact with the machine you have deployed
+- It always runs on the machine's private network ip and port number equla to `20000 +last octect` of machine private IP
+- For exampleif the machine ip is `10.20.2.2/24` this means
```
Comment (xmonader) [inline docs/manual/zmachine/cloud-console.md]: for example, if the

**Review by xmonader — DISMISSED** [review]

File: pkg/vm/ch.go [hunk 14]
```diff
@@ -17,9 +19,43 @@ import (
 )
 
 const (
-	chBin = "cloud-hypervisor"
+	chBin           = "cloud-hypervisor"
+	cloudConsoleBin = "cloud-console"
 )
 
+// StartCloudConsole Starts the cloud console for the vm on it's private network ip
+func (m *Machine) StartCloudConsole(ctx context.Context, namespace string, networkAddr net.IPNet, machineIP net.IPNet, ptyPath string) error {
```
Comment (muhamadazmy) [inline pkg/vm/ch.go]: Why this function is defined on a Machine receiver not on the Manager. It does not access any of the Machine properties.

I think it's better for this to become a `private` manager method instead it fits more there.

File: pkg/vm/client.go [hunk 15]
```diff
@@ -86,40 +93,48 @@ func (c *Client) Resume(ctx context.Context) error {
...
+		body, _ := ioutil.ReadAll(response.Body)
+		response.Body.Close()
+
+		return VMData{}, fmt.Errorf("got unexpected http code '%s' on machine info, Response: %s", response.Status, string(body))
 	}
 
 	var data struct {
 		Config struct {
-			CPUs struct {
+			CPU struct {
```
Comment (muhamadazmy) [inline pkg/vm/client.go:119]: Why not use the `VMData` struct here instead of anonymous  struct then copy everything. Just make sure VMData has the right json annotiations then return data.Config directly

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

**Comment by muhamadazmy** [comment]
BLOCKED ON: https://github.com/LeeSmet/cloud-console/issues/3

**Review by muhamadazmy — APPROVED** [review]
